OpenRA.Mods.Common/Traits/Player/TimeLimitManager.cs

```diff
@@ -0,0 +1,165 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Mods.Common.Widgets;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+using OpenRA.Widgets;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("This trait allows setting a time limit on matches. Attach this to the World actor.")]
+	public class TimeLimitManagerInfo : ITraitInfo, ILobbyOptions
+	{
+		[Desc("Label that will be shown for the time limit option in the lobby.")]
+		public readonly string TimeLimitLabel = "Time Limit";
+
+		[Desc("Tooltip description that will be shown for the time limit option in the lobby.")]
+		public readonly string TimeLimitDescription = "Time Limit before a match ends automatically.";
+
+		[Desc("Time Limit options that will be shown in the lobby dropdown. Values are in minutes.")]
+		public readonly int[] TimeLimitOptions = { 0, 10, 20, 30, 40, 60, 90 };
+
+		[Desc("List of remaining minutes of game time when a text and optional speech notification should be made to players.")]
+		public readonly Dictionary<int, string> TimeLimitWarnings = new Dictionary<int, string>
+		{
+			{ 1, null },
+			{ 2, null },
+			{ 3, null },
+			{ 4, null },
+			{ 5, null },
+			{ 10, null },
+		};
+
+		[Desc("Default selection for the time limit option in the lobby. Should use one of the TimeLimitOptions.")]
+		public readonly int TimeLimitDefault = 0;
+
+		[Desc("Prevent the time limit option from being changed in the lobby.")]
+		public readonly bool TimeLimitLocked = false;
+
+		[Desc("Display order for the time limit dropdown in the lobby.")]
+		public readonly int TimeLimitDisplayOrder = 0;
+
+		[Desc("Notification text for time limit warnings. The string '{0}' will be replaced by the remaining time in minutes, '{1}' is used for the plural form.")]
+		public readonly string Notification = "Match timer: {0} minute{1} remaining.";
+
+		[Desc("Label used to display a text ingame that will be updated every second.")]
+		public readonly string CountdownLabel = null;
+
+		[Desc("Text to be shown using the CountdownLabel. The string '{0}' will be replaced by the time in hh:mm:ss format.")]
+		public readonly string CountdownText = null;
+
+		[Desc("Will prevent showing/playing the built-in time limit warnings when set to true.")]
+		public readonly bool SkipTimeRemainingNotifications = false;
+
+		[Desc("Will prevent showing/playing the built-in timer expired notification when set to true.")]
+		public readonly bool SkipTimerExpiredNotification = false;
```

Comment:
Not currently, but it could be used for game modes that used waves or phases or whatever people might come up with.

I don't really want to move it to *VictoryConditions as those are per player.